File: 单调栈/1. 每日温度 - 739/Daily Temperatures.cpp
/*
    23/03/24
    https://leetcode.cn/problems/daily-temperatures/
 */

#include <iostream>
#include <vector>
#include <stack>

using namespace std;

/*
    solution: 栈
    result: ✅
    TC: O(n)
    SC: O(1)
 */

vector<int> dailyTemperatures(vector<int>& temperatures) {
    
    int length = temperatures.size();
    vector<int> result(length, 0);
    stack<int> stk;

    for (int cur = 0; cur < length; cur ++) {
        while (!stk.empty() && temperatures[cur] > temperatures[stk.top()]) {
            auto tempIndex = stk.top();
            stk.pop();
            result[tempIndex] = cur - tempIndex;
        }
        stk.push(cur);
    }

    return result;
}

File: 单调栈/1. 每日温度 - 739/Daily Temperatures.py
# 23/03/24
# https://leetcode.cn/problems/daily-temperatures/

from collections import List

def dailyTemperatures(temperatures: List[int]) -> List[int]:

    length = len(temperatures)
    result = [0] * length
    stack = list()

    for cur in range(length):
        while stack and temperatures[cur] > temperatures[stack[-1]]:
            tempIndex = stack.pop()
            result[tempIndex] = cur - tempIndex
        stack.append(cur)
    
    return result

dailyTemperatures([73,74,75,71,69,72,76,73]) # [1,1,4,2,1,1,0,0]
dailyTemperatures([30,40,50,60]) # [1,1,1,0]
dailyTemperatures([30,60,90]) # [1,1,0]